### PR TITLE
[Training] Refactor loss functions and utils

### PR DESCRIPTION
**Pull request by Ubospica**
This PR refactors the loss interface and `AppendLoss` utility to better support state parameters and return values in the backbone.

The new structure of loss and AppendLoss:

```
@R.function
def backbone(input_of_backbones, input_of_states):
    # You can return the updated model states in `output_of_states`, such as `running_mean` and `running_var` in `batch_norm`
    return prediction_outputs, output_of_states

@R.function
def loss(backbone_prediction_outputs, targets):
    return loss

---> After pass:

@R.function
def backbone_loss(input_of_backbones, input_of_states, targets):
    # all bindings in backbone and loss
    return loss, output_of_states
```

Also we have added a `num_backbone_outputs` attribute for `Loss` objects. It denotes the length of the `prediction_outputs` return value of the backbone function, as well as the `backbone_prediction_outputs` parameter of loss function. 

`num_backbone_outputs` may be different for various Losses. It can be specified when defining the Loss class. 

This PR also makes these changes:

- Change AppendLoss to a pass. Doing so can make our `SetupTrainer` a combination of passes.
```
def AppendLoss(
    func_name: str,
    loss_function: Function,
    num_backbone_outputs: int = 1,
    new_func_name: Optional[str] = None,
) -> tvm.ir.transform.Pass:
    pass
```
- Refactor AppendLoss thoroughly to clarify the logic
- Polish a lot of docs

### CODE REVIEW

File: tests/python/relax/test_transform_gradient.py [hunk 1]
```diff
@@ -99,7 +98,7 @@ def main_adjoint(x: R.Tensor((3, 3), "float32")) -> R.Tuple(R.Tensor((), "float3
     # fmt: on
 
     After = relax.transform.Gradient("main")(Before)
-    assert_structural_equal(After["main_adjoint"], Expected["main_adjoint"])
+    assert_structural_equal(After, Expected)
```
Comment (Ubospica) [inline tests/python/relax/test_transform_gradient.py:101]: Also changes a few Gradient unit tests

**Review by SiriusNEO — COMMENTED** [review]
Given that the usage of `AppendLoss` is not so general (It is specially designed for loss), is it appropriate that we put it in `transform`? Should we move it to the `training` namespace?

**Review by SiriusNEO — APPROVED** [review]
